pkg/cli/tui/spinner.go

```diff
@@ -0,0 +1,255 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tui provides UI elements for the CLI.
+package tui
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/charmbracelet/bubbles/help"
+	"github.com/charmbracelet/bubbles/key"
+	"github.com/charmbracelet/bubbles/spinner"
+	tea "github.com/charmbracelet/bubbletea"
+	"go.uber.org/zap"
+	"golang.org/x/term"
+)
+
+const (
+	spinnerInterval = 150 * time.Millisecond
+)
+
+var (
+	currentStepTitleStyle = helperTextStyle.Bold(false)
+	successStepTitleStyle = successStyle
+	failedStepTitleStyle  = failureStyle
+	successMark           = successStyle.Render("✅")
+	failureMark           = failureStyle.Render("❌") // x
+)
+
+type (
+	// Step provides a way to run a function with a
+	// pretty loading spinner animation.
+	Step struct {
+		// Desc is a human-readable description of the step.
+		Desc string
+		// F is the function that will be called to execute the step.
+		F func(ctx context.Context) error
+	}
+
+	// spinnerKeyMap defines a set of keybindings. To work for help it must satisfy
+	// key.Map. It could also very easily be a map[string]key.Binding.
+	spinnerKeyMap struct {
+		Quit key.Binding
+	}
+
+	// Spinner represents action with spinner element.
+	Spinner struct {
+		steps     []Step
+		index     int
+		keys      spinnerKeyMap
+		help      help.Model
+		spinner   spinner.Model
+		p         *tea.Program
+		interrupt bool               // set in case user wants to quit (Esc or Ctrl+c)
+		done      bool               // all steps have been completed
+		stepError error              // error occurred during step execution
+		ctx       context.Context    // upper lavel context
+		l         *zap.SugaredLogger // logger
+	}
+
+	// SpinnerOption is used to set options when initializing Spinner.
+	// Spinner can accept a variable number of options.
+	//
+	// Example usage:
+	//
+	//	p := NewSpinner(ctx, WithPretty(bool))
+	SpinnerOption func(m *Spinner)
+)
+
+// ShortHelp returns keybindings to be shown in the mini help view. It's part
+// of the key.Map interface.
+func (k spinnerKeyMap) ShortHelp() []key.Binding {
+	return []key.Binding{k.Quit}
+}
+
+// FullHelp returns keybindings for the expanded help view. It's part of the
+// key.Map interface.
+func (k spinnerKeyMap) FullHelp() [][]key.Binding {
+	return [][]key.Binding{
+		{k.Quit}, // first column
+		{},       // second column
+	}
+}
+
+// NewSpinner creates a new spinner element.
+func NewSpinner(ctx context.Context, l *zap.SugaredLogger, steps []Step, opts ...SpinnerOption) Spinner {
+	s := spinner.New(
+		spinner.WithSpinner(spinner.Points),
```

Comment:
What's the reason for changing it from line to dots? I personally liked the line animation more because it was a bit more subtle 😅

Reply:
Well, new lib is used - new styles were proposed. It was reviewed by Nina - she approved.